Подписка на других пользователей

nurlantr opened the pull request.

nurlantr committed:
Подписка на других пользователей

egorbarzenkov commented on src/main/java/school/faang/user_service/service/SubscriptionService.java:
```diff
@@ -0,0 +1,69 @@
+        subscriptionRepository.followUser(followerId, followeeId);
+    }
+
+    public void unfollowUser(long followerId, long followeeId) {
```
лучше возвращать в контроллер успешно ли прошла операция или нет

egorbarzenkov commented on src/main/java/school/faang/user_service/controller/SubscriptionController.java:
```diff
@@ -0,0 +1,46 @@
+
+import java.util.List;
+
+@Component
```
переделай на контроллер, это уже тема текущего спринта

egorbarzenkov commented on src/main/java/school/faang/user_service/controller/SubscriptionController.java:
```diff
@@ -0,0 +1,46 @@
+public class SubscriptionController {
+    private final SubscriptionService subscriptionService;
+
+    public void followUser(long followerId, long followeeId) throws DataValidationException {
```
лучше возвращать состояние операции, успешно она прошла или произошла какая то ошибка

egorbarzenkov commented on src/main/java/school/faang/user_service/service/SubscriptionService.java:
```diff
@@ -0,0 +1,69 @@
+
+    public List<User> filterUsers(List<User> users, UserFilterDto filter) {
+        return users.stream()
+                .filter(user -> filter.getNamePattern() == null || user.getUsername().contains(filter.getNamePattern()))
```
лучше делать это через reduce, можно пробегать по списку фильтров и проверять подходит ли он для применения к фильтру и если подходит то уже применять его фильтрацию
так же сделай фильтры как компоненты спринга, чтобы была возможность за счет него управлять поведением фильтрации

egorbarzenkov commented on src/test/java/school/faang/user_service/service/service/SubscriptionServiceTest.java:
```diff
@@ -0,0 +1,196 @@
+    @Mock
+    private SubscriptionRepository subscriptionRepository;
+
+    @Mock
```
для компонентов с примитивной логикой не обязательно использовать @Mock, лучше подойдет @Spy

egorbarzenkov commented on src/main/java/school/faang/user_service/dto/UserDto.java:
```diff
@@ -0,0 +1,12 @@
+package school.faang.user_service.dto;
```
создай подпакет user и в него закинь свои сущности

nurlantr committed:
Изменения:
1) в Controller класс добавил все что прошли за это время и обновил ControllerTest класс
2) Перенес UserDto и UserFilterDto в под-папку этой сущности
3) Обновил класс Service чтобы каждый метод возвращал в контроллер успешно ли прошла операция или нет
4) Так же в Service классе обновил метод filterUsers и для этого создал новую папку (filter) со всеми нужными классами для фильтров.
5) Затем соответственно обновил ServiceTest класс
6) Изменил аннотацию @Mock на @Spy где есть примитивная логика

egorbarzenkov commented on src/main/java/school/faang/user_service/controller/SubscriptionController.java:
```diff
@@ -0,0 +1,71 @@
+import java.util.List;
+
+@RestController
+@RequiredArgsConstructor
```
укажи через RequestMapping доменную область, чтобы было явно, с чем работает контроллер

egorbarzenkov commented on src/main/java/school/faang/user_service/controller/SubscriptionController.java:
```diff
@@ -0,0 +1,71 @@
+    public ResponseEntity<String> followUser(@RequestParam long followerId, @RequestParam long followeeId) {
+        try {
+            if (followerId == followeeId) {
+                throw new DataValidationException("You cannot follow yourself!");
```
затяни к себе ветку murloc-master-bc5 и вынеси наименования всех своих ошибок в класс ExceptionMessages

egorbarzenkov commented on src/main/java/school/faang/user_service/service/SubscriptionService.java:
```diff
@@ -0,0 +1,67 @@
+                .toList();
+    }
+
+    public List<User> filterUsers(List<User> users, UserFilterDto filter) {
```
почему метод публичный, если он используется только внутри сервиса?

egorbarzenkov commented on src/main/java/school/faang/user_service/mapper/UserMapper.java:
```diff
@@ -0,0 +1,12 @@
+import school.faang.user_service.entity.User;
+
+@Mapper(componentModel = "spring", unmappedTargetPolicy = ReportingPolicy.IGNORE)
+public interface UserMapper {
```
после вливания ветки murloc-master-bc5 в свою ветку оставь одну сущность UserDto в проекте

egorbarzenkov commented on src/main/java/school/faang/user_service/filter/user/UserFilter.java:
```diff
@@ -0,0 +1,8 @@
+import school.faang.user_service.dto.user.UserFilterDto;
+import school.faang.user_service.entity.User;
+
+public interface UserFilter {
```
используй уже имеющуюся реализацию UserFilter из ветки murloc-master-bc5, соответственно тебе нужно будет переписать логику работы в сервисе, но это будет уже не так сложно, подобные реализации есть

nurlantr replied:
там только есть для двух фильтров. У меня их 10) Мне тогда просто по логике двух фильтров остальные написать, а свою папку с фильтрами удалить?

egorbarzenkov replied:
фильтр по юзерам можно расширять завязавшись на UserFilterDto, по этому докинь свои фильтры к уже имеющимся и их выполнение или не выполнение можно будет контролировать на уровне методы проверки применения к UserFilterDto

egorbarzenkov commented on src/main/java/school/faang/user_service/controller/SubscriptionController.java:
```diff
@@ -0,0 +1,71 @@
+public class SubscriptionController {
+    private final SubscriptionService subscriptionService;
+
+    @PostMapping("/follow")
```
Добавь value из аннотаций в ApiPath из ветки murloc-master-bc5, чтобы они были в одном месте и не приходилось искать path по проекту

nurlantr replied:
мне для каждого маппинга это делать? или только который на класс вешается (@RequestMapping)?

egorbarzenkov replied:
для каждого, но обрати внимание, что /followers/count можно разбить на 
/followers + /count тем самым можно будет выделить меньше наименования для переменных

nurlantr committed:
Merge branch 'murloc-master-bc5' into feature-BJS2-15750

# Conflicts:
#	src/main/java/school/faang/user_service/exception/DataValidationException.java
#	src/main/java/school/faang/user_service/mapper/UserMapper.java

nurlantr committed:
Изменения:
1) в Controller класс добавил доменную область и обновил ApiPath
2) Обновил ExceptionMessages
3) Переименовал UserFilter на UserSubscriptionFilter и поменял его интерфейс и соответственно обновил все реализации (10 классов)
4) Обновил SubscriptionService так как реализации фильтров изменились. Так же обновил SubscriptionServiceTest

nurlantr committed:
.
